### PR TITLE
Implement all commands of Ctap2

### PR DESCRIPTION
**Pull request by msirringhaus**
These are now all the commands defined in the spec.
Not all functionality is there yet, and a few testcases are missing, as the spec doesn't give examples for those.
Most notably: All crypto-usage is still missing. For this, we need something like https://github.com/mozilla/authenticator-rs/pull/142 with more functionality and then implement that for different crypto backends (NSS, openssl, ring, ..).

Next step would be to create a new Manager and StateMachine, which uses the new commands.

### CODE REVIEW

**Comment by bertvandepoel** [comment]
I don't really know much about the spec, so I was wondering whether the crypto-usage stuff, which requires #142 to be finalised and implemented, is a strict requirement or a "nice to have". I assume this PR can't be merged before the manager and statemachine are done and there are no more WIP commits then? Mostly just trying to understand whether webauthn support in firefox is in the near future or not at all.

**Comment by msirringhaus** [comment]
The crypto-stuff is a strict requirement for hardware tokens that support a PIN (which is pretty much most of the FIDO2 compatible ones). In my branch, I already have an openssl-backend going (probably highly insecure) which works, and am currently working on the NSS-backend (need to check for existing bindings or write my own). 

The statemachine and manager are mostly functional (although some CTAP1-backwards compatibility is still missing). 
Those are all not in the PR yet, as it is pretty huge already and I don't want to make it unreviewable. 

Once all of that is done (assuming it passes review at some point), we still need to integrate it into Firefox (Rust<->C glue, new User dialogs, etc.). So yes, its coming, but not tomorrow. I still work more or less full time on this.

Just look at my fork on the `ctap2-cont` branch for the current state. 

In case you (or @theSuess) want to help: More testcases would be nice. Esp. for the GetAssertion and GetNextAssertion commands.

**Comment by bertvandepoel** [comment]
I'm coming back to this after a busy summer (sorry for the delay!), and was wondering what you exactly mean with testcases? Do you mean use our devices to test things, or more literally unit tests. The first one I could probably help with with my Mooltipass miniBLE, but the latter might be a bit out of my league since I don't know any Rust. I see @theSuess wrote https://github.com/msirringhaus/authenticator-rs/pull/1 but I'm not too sure whether or not that's related.

**Comment by msirringhaus** [comment]
Now its my turn to apologize for the delay :-) 
In my original statement, I actually meant unit tests, yes (exactly like the one you mentioned). But trying out as many devices as possible would also be very beneficial! 

And thank you for pointing out @theSuess 's PR to me, as I completely missed that one (github somehow didn't notify me about it).

As a quick general update: This project has been on hold for the past two months or so, because of unrelated FF91esr-work (plus some more weeks of vacation), which is still ongoing. But hopefully, this will be done soon, so I can pick this up again. 
In theory, I now have a NSS crypto-backend going, and was able to put my code (locally) into Firefox and successfully use CTAP2 on https://demo.yubico.com/ . Most of the Firefox-integration C++ code is still very hacky and it is missing a PIN-dialog for the user,  but I personally would consider the Rust part "almost good enough for now". But all of this is needs to be reviewed and accepted by Mozilla's security team, of course. They probably will have a lot to say about "good enough", I'm sure ;-)

**Comment by bertvandepoel** [comment]
No worries, I presumed some vacation was for sure involved. Important to recharge your batteries. If you at some stage would like em to test with my Mooltipass miniBLE, then feel free to let me know!

File: src/ctap2/commands/mod.rs
Comment (rmartinho) [inline src/ctap2/commands/mod.rs:328]: Can we name these two consistently? Either Deerializing/Serializing, or Deserialization/Serialization

File: src/ctap2/commands/make_credentials.rs
Comment (rmartinho) [inline src/ctap2/commands/make_credentials.rs:213]: These `map_err` calls are redundant, right? Can we remove them and just return Retryable::Error in the first one?

File: src/ctap2/commands/get_assertion.rs
Comment (rmartinho) [inline src/ctap2/commands/get_assertion.rs:275]: These map_err calls are redundant, right? Can we just return `Retryable::Error`?

File: src/ctap2/commands/get_assertion.rs
Comment (rmartinho) [inline src/ctap2/commands/get_assertion.rs:325]: Same as here as elsewhere, why the redundant map_err?

I won't comment on all instances to not be spammy, but this kind of redundant map_err calls occur in other places in this and other files.

**Review by rmartinho — COMMENTED** [review]
Ignoring all the TODO stuff, this looks good so far, I just left a few suggestions here and there.

File: Cargo.toml
Comment (msirringhaus) [reply Cargo.toml:57]: I have tried  to use the versions that are already vendored in Firefox, to avoid update procedures there that might cause problems. Currently, Firefox has sha2 in version 0.8.2. 
If you think it's not a problem to bump that dependency in Firefox as well, I can go with 0.9.

File: src/ctap2/client_data.rs
Comment (msirringhaus) [reply src/ctap2/client_data.rs:126]: This is already fixed in my follow-up branch. It is not an Option anymore, and serde is using bool::default now.
Do you want me to backport that, or just look at it with the next PR?

**Comment by msirringhaus** [comment]
That would be it from my side for now. Remaining questions are `sha2`-version bump and backporting of the crossOrigin-changes. I hope the follow-up commits resolve the other issues.

File: src/ctap2/client_data.rs
Comment (rmartinho) [reply src/ctap2/client_data.rs:126]: That's ok, we can look at it in the next one, then.

**Review by rmartinho — APPROVED** [review]
This looks fine as is. The sha2 version bump we can handle later, and there's no need to spend time backporting stuff onto this if they're coming in the follow-up PR anyway.

**Pull request merged**
